crates/binjs_io/src/entropy/predict.rs

```diff
@@ -336,3 +346,172 @@ impl<NodeValue> PathPredict<NodeValue, SymbolInfo> where NodeValue: Eq + Hash +
     }
 }
 
+
+
+
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+enum WindowPrediction {
+    BackReference(BackReference),
+    DictionaryIndex(DictionaryIndex),
+}
+
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct WindowPredict<NodeValue, Statistics> where NodeValue: Clone + Eq + Hash {
+    /// The window width
+    width: usize,
+
+    /// All the known values.
+    value_by_dictionary_index: Vec<NodeValue>,
+    dictionary_index_by_value: HashMap<NodeValue, DictionaryIndex>,
+
+    latest_values: Vec<NodeValue>,
+
+    info: ContextInformation<WindowPrediction, Statistics>,
+}
+impl<NodeValue, Statistics> WindowPredict<NodeValue, Statistics> where NodeValue: Clone + Eq + Hash {
+    pub fn new(width: usize) -> Self {
+        WindowPredict {
+            width,
+            value_by_dictionary_index: Vec::with_capacity(1024), // FIXME: Magic number
+            dictionary_index_by_value: HashMap::with_capacity(1024),
+            latest_values: Vec::with_capacity(width),
+            info: ContextInformation::new(),
+        }
+    }
+
+    /// Update current window by moving an index to the first position.
+    fn update_window_by_index(&mut self, index: BackReference) -> Result<BackReference, ()> {
```

Comment:
the method name doesn't sound correct.
what this does is, remove the entries after index-th value, and move the index-th value to 0-th element.
so, at least "by" is misleading (for me it sounds like "+=")

also, what's the reason why the entries after index-th value is dropped?

Reply:
Not quite. Method `as_mut_slice` gives a view on a subset of the vector, so we're not dropping anything. In that subset, we replace the last value, then rotate right.